### DISCUSSION TITLE
How to only save instance of a class once Checkout Fulfilled (One-time payment)

### DISCUSSION THREAD

**Discussion by SuDa2103** [pay-rails/pay#517] [Q&A] [answered]
Hi all, 

Loving this gem! Just a quick question. 

Let's say I have a Job model. And I want to make sure the user has paid before the Job they have created is saved, does anyone know how I can do this with the pay gem? I've been trying to figure out how to tweak the code below (which is in the docs) to do this. 

```
# config/initializer/pay.rb
class FulfillCheckout
    def call(event)
      object = event.data.object
      return object.payment_status != "paid"
      # Handle fulfillment
    end
  end
```

The workaround I was thinking of was adding a paid column to the Jobs table, which would be a boolean, and then set it to true when checkout is complete. Then I could only show the Jobs where paid == true

**Comment by cjavilla-stripe** [comment, marked as answer]
Since the webhook is asynchronous, the approach I would take is track the `status` of the Job instead of waiting to save it. 

1. Create and save the Job with `status` = `:pending`
2. Send customer through Checkout
3. In FulfillCheckout, if the `payment_status` of the Checkout Session is paid, update the `status` = `:paid`

**Reply by SuDa2103** [reply]
Hi There!
Thanks for this @cjavilla-stripe , and sorry to bother again, but I'm trying to do something similar to this, and finding that my code for FulFillCheckout is never being reached. But my webhooks are working locally. Are there any common mistakes with the FulFillCheckout code? I have defined the class above and have these two lines in my pay initializer:

Pay::Webhooks.delegator.subscribe "stripe.checkout.session.completed", FulfillCheckout.new
Pay::Webhooks.delegator.subscribe "stripe.checkout.session.async_payment_succeeded", FulfillCheckout.new

**Reply by cjavilla-stripe** [reply]
I ran into this too. I defined my `FulfillCheckout` thing _inside_ the initializer for Pay so that I knew for sure that they were loaded. I used a lambda instead of a class, but they should work the same. Here's what my initializer looks like:

```rb
# config/initializers/pay.rb
Pay.setup do |config|
  # For use in the receipt/refund/renewal mailers
  config.business_name = "Business Name"
  config.business_address = "1600 Pennsylvania Avenue NW"
  config.application_name = "My App"
  config.support_email = "helpme@example.com"

  config.send_emails = true

  config.default_product_name = "default"
  config.default_plan_name = "default"

  config.automount_routes = true
  config.routes_path = "/pay" # Only when automount_routes is true
end

# This could also be where you define the class with `class FulfillCheckout`
fulfill_checkout = -> (event) {
  checkout_session = event.data.object
  # Here's where you'd refetch the session and expand it's line_items:
  # checkout_session = Stripe::Checkout::Session.retrieve({
  #   checkout_session.id,
  #   expand: ['line_items']
  # })
  customer = Pay::Customer.find_by(processor_id: checkout_session.customer)
  customer.owner.clear_cart!

  return if checkout_session.payment_status != 'paid'

  # Pull from inventory
  # Print shipping labels
  # Notify in chat
}

Pay::Webhooks.delegator.subscribe "stripe.checkout.session.async_payment_succeeded", fulfill_checkout
Pay::Webhooks.delegator.subscribe "stripe.checkout.session.completed", fulfill_checkout

```

**Reply by SuDa2103** [reply]
Thanks for this @cjavilla-stripe! This is working for me now

**Reply by felipecerda** [reply]
Thanks @cjavilla-stripe, that comment would be a great addition to the pay documentation to make that fulfillment section a bit more complete.
